synapse/federation/sender/__init__.py

```diff
@@ -393,34 +420,59 @@ async def handle_room_events(events: Iterable[EventBase]) -> None:
         finally:
             self._is_processing = False
 
-    async def _send_pdu(self, pdu: EventBase, destinations: Iterable[str]) -> None:
-        # We loop through all destinations to see whether we already have
-        # a transaction in progress. If we do, stick it in the pending_pdus
-        # table and we'll get back to it later.
+    async def _distribute_events(
+        self,
+        per_room_events_and_dests: List[
+            Tuple[str, List[Tuple[EventBase, Collection[str]]]]
+        ],
+    ) -> None:
+        # Tuples of room_id + destination to their max-seen stream_ordering
+        room_with_dest_stream_ordering = {}  # type: Dict[Tuple[str, str], int]
 
-        destinations = set(destinations)
-        destinations.discard(self.server_name)
-        logger.debug("Sending to: %s", str(destinations))
+        # List of events to send to each destination
+        events_by_dest = {}  # type: Dict[str, List[EventBase]]
 
-        if not destinations:
-            return
+        # For each roomID + events with destinations pair...
+        for room_id, event_and_dests in per_room_events_and_dests:
+
+            # (dest -> stream_id)
+            this_stream_ordering = {}  # type: Dict[str, int]
+
+            # ...get every event with their destinations...
+            for event, destinations in event_and_dests:
 
-        sent_pdus_destination_dist_total.inc(len(destinations))
-        sent_pdus_destination_dist_count.inc()
+                # (we got this from the database, it's filled)
+                assert event.internal_metadata.stream_ordering
 
-        assert pdu.internal_metadata.stream_ordering
+                sent_pdus_destination_dist_total.inc(len(destinations))
+                sent_pdus_destination_dist_count.inc()
 
-        # track the fact that we have a PDU for these destinations,
-        # to allow us to perform catch-up later on if the remote is unreachable
-        # for a while.
-        await self.store.store_destination_rooms_entries(
-            destinations,
-            pdu.room_id,
-            pdu.internal_metadata.stream_ordering,
+                # ...iterate over those destinations..
+                for destination in destinations:
+                    if (
+                        this_stream_ordering.setdefault(destination, 0)
+                        < event.internal_metadata.stream_ordering
+                    ):
+                        # ...update their stream-ordering...
+                        this_stream_ordering[
+                            destination
+                        ] = event.internal_metadata.stream_ordering
```

Comment:
also: can't we assume that we receive events in `stream_ordering` order, so we can make this unconditional? I'm pretty sure that other things depend on this.

Reply:
I dont want to assume we receive events in `stream_ordering`, because i have not seen that we *do*, regardless, i think this has minimal performance impact for a good fallback "collection", which can assure more redundancy.

I'll apply the suggestion above :+1: